### PR TITLE
Fix clobbering of overridden canvas methods

### LINKED ISSUE #1226: Chrome Extension clashing with HiDPI canvas polyfill
HiDPI canvas polyfill is really famous and great library for HiDPI retina canvas support (if you don't use it you get blured canvas on retina/HiDPI screens)

Here is the explanation of what lib does:
http://jondavidjohn.github.io/hidpi-canvas-polyfill/

Here is the issue on their github:
https://github.com/jondavidjohn/hidpi-canvas-polyfill/issues/29
 

I know many people that use this in their production code, it would be such a shame for them to experience this just by having this great extension, such the Privacy Badger is.

Chrome Version: 56.0.2924.87 (64-bit)

Please advise!

### LINKED ISSUE #1854: Privacy Badger causing rendering issues with charts
PrivacyBadger is causing rendering issues on Tradingview.com with Chrome version 64.0.3282.119 (Windows & Chrome OS). This occurs when "Disable Privacy Badger for This Site" as well as if the site is explicitly whitelisted in PB's options. 

Example of Privacy Badger running ( includes while whitelisted or disabled):
image

Console output with Privacy Badger running:
image

Example of Privacy Badger removed:
image

Possibly related: https://github.com/EFForg/privacybadger/issues/1820

### LINKED ISSUE #2657: Privacy Badger breaks text on Canvas charting software
#### What is your browser and browser version?
Chrome 84.0.4147.89, MacOS, 2020 Macbook (HiDPI)

#### What is broken and where?
The chart at https://www.bitmex.com/app/trade/XBTUSD ([full screen](https://www.bitmex.com/chartEmbed?symbol=XBTUSD)) has broken text labels and a broken X axis. They appear to be rendering at half width/height, suggesting Privacy Badger is breaking HiDPI somehow.

With Privacy Badger enabled, notice incorrect render output on X axis and on text labels on both axes:

Screen Shot 2020-07-30 at 10 16 20 AM

With it disabled, notice correct output and readable text on labels and X axis:

Screen Shot 2020-07-30 at 10 17 50 AM

#### What is the "culprit" domain?
No trackers blocked. I believe something the canvas fingerprinting protection is doing is causing this breakage.

#### What is your debug output for this domain?
No debug output.

Possibly related to #1226

### PR DESCRIPTION
**Pull request by STRML**
```
The method may be overridden again after we override,
as is done in the `hidpi-canvas-polyfill`. So if we
restore the method to the original as we see it, we'll
wipe out the polyfill.
```

I'd like to write tests for this, but I see very little related to canvas fingerprinting, and what is there is pretty inscrutable (where is `fingerprinting.html`?)

Fixes #2657, fixes #1226, fixes #1854.

### CODE REVIEW

**Comment by ghostwords** [comment]
Hello, and thank you for the report and the PR! I will take a proper look next week.

[`fingerprinting.html`](https://github.com/EFForg/privacybadger/blob/0a8a9a7a7bfc34bd5d2385d1df9f60f3987ef6cb/tests/selenium/fingerprinting_test.py#L44) lives in https://github.com/EFForg/privacybadger-test-fixtures. (We should probably document where fixtures live [somewhere in our tests doc](https://github.com/EFForg/privacybadger/blob/master/doc/tests.md#functional-tests).)

Let me know what other questions you have. Thanks again!

**Comment by STRML** [comment]
Thanks @ghostwords. Perhaps this should be same-repo, and we just run a local webserver rather than hit the github.io page? Adding or modifying a test requires now coordinating two PRs, which is quite difficult. Regardless, I'll make a PR there that correlates to this one, but it will need to be accepted and the tests re-run here.

**Comment by ghostwords** [comment]
Yeah, the coordination is a bit of a pain. We used to host test fixtures on various public JS playground/gist hosts, but for various reasons (reliability, control over what actually gets served) migrated to the two GH Pages repos we have now. #928 has a bit of history and links to some relevant commits.

A local fixtures server would be nice, but we need support for things like https and distinct-to-Privacy Badger hostnames (to simulate first and third-party interactions). (Localhost domains are probably not a good fit since they should probably be ignored: #817)

**Comment by STRML** [comment]
That's fair enough. So we'd have to run a system-level dns resolver and point the local browsers to it... quite a song & dance. Doubt it's worth the effort.

**Comment by STRML** [comment]
Build failing, I'll ping when this is ready for merge.

**Comment by ghostwords** [comment]
I should have some time to look into what's going on with the text next week, if that's helpful.

**Comment by STRML** [comment]
It would be, thanks. When running locally I have no issue.

**Comment by ghostwords** [comment]
I think the test doesn't work on Travis because it doesn't run in a HiDPI environment. We could fake it with something like `window.devicePixelRatio = 2;`. Still looking though.

**Comment by ghostwords** [comment]
Looks like that was it, the polyfill wasn't kicking in on Travis. I added the `devicePixelRatio` override, the test should pass now.

File: src/js/contentscripts/fingerprinting.js [hunk 1]
```diff
@@ -204,14 +204,19 @@ function getFpPageScript() {
       );
 
       item.obj[item.propName] = (function (orig) {
+        // set to true after the first write, if the method is not
+        // restorable. Happens if another library also overwrites
+        // this method.
+        var skipMonitoring = false;
```
Comment (ghostwords) [inline src/js/contentscripts/fingerprinting.js:210]:
```javascript
let skip_monitoring = false;
```

File: tests/selenium/fingerprinting_canvas_hidpi_test.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# -*- coding: UTF-8 -*-
+
+import unittest
+
+import pbtest
+
+from pbtest import retry_until
+
+
+class FingerprintingTest(pbtest.PBSeleniumTest):
```
[the remark this answers is not in the document]
Comment (STRML) [reply tests/selenium/fingerprinting_canvas_hidpi_test.py:11]: It requires a different page, so I figured it should be a different class, but I'm fine with putting it in there.

**Comment by STRML** [comment]
@ghostwords We're green, thanks a lot for the help and for catching the fixture issue, I wouldn't have caught that.

I can clean up the history if you like, or if you're just going to squash it should be no problem.

**Comment by ghostwords** [comment]
Thank you for the contribution! I'll finish testing next week, expect to merge then in time for the next Privacy Badger update.

I won't squash merge as I would rather preserve the individual commits. You're welcome to clean up history, but it's not necessary.

**Comment by ghostwords** [comment]
Thanks again!

**Pull request merged**